lib/SILOptimizer/Mandatory/DataflowDiagnostics.cpp

```diff
@@ -117,35 +119,62 @@ static void diagnoseStaticReports(const SILInstruction *I,
 // SWIFT_ENABLE_TENSORFLOW
 /// \brief Emit a diagnostic for `poundAssert` builtins whose condition is
 /// false or whose condition cannot be evaluated.
-/// TODO(clattner / marcrasi): Implement this for real.
-static void diagnosePoundAssert(const SILInstruction *I, SILModule &M) {
+static void diagnosePoundAssert(const SILInstruction *I, SILModule &M,
+                                ConstExprEvaluator &constantEvaluator) {
   if (auto *builtinInst = dyn_cast<BuiltinInst>(I)) {
     if (builtinInst->getBuiltinKind() == BuiltinValueKind::PoundAssert) {
-      diagnose(M.getASTContext(), I->getLoc().getSourceLoc(),
-               diag::pound_assert_not_implemented);
+      SmallVector<SymbolicValue, 1> values;
+      constantEvaluator.computeConstantValues({builtinInst->getArguments()[0]},
+                                              values);
+      SymbolicValue value = values[0];
+      if (!value.isConstant()) {
+        // TODO(marcrasi / clattner): Emit more informative error.
+        diagnose(M.getASTContext(), I->getLoc().getSourceLoc(),
+                 diag::pound_assert_condition_not_constant);
+        return;
+      }
+      if (value.getTypeKind() != SymbolicValue::TKInteger) {
+        llvm_unreachable("sema prevents non-integer #assert condition");
+      }
+
+      APInt intValue = value.getIntegerValue();
+      if (!intValue.isSignedIntN(1)) {
```

Comment:
Similarly, use an assertion.  A better spelling is: assert(intValue.getBitWidth() == 1 && "...")